File: 0011-container-with-most-water/0011-container-with-most-water.cpp
class Solution {
public:

    int maxArea(vector<int>& height) {
        int count = 0;
        int maxVolume = 0;
        //int distance = height.size();
        int rightMovement = height.size()-1;
        int leftMovement = 0;


        int left = height[0];
        int right = height[height.size()-1];

        for (int i = 0; i < height.size(); i++) {
            left = height[i];
             leftMovement = i;

            maxVolume = max(maxVolume, min(left, right) * (rightMovement - leftMovement));

            if (right < left) {
                for (int j = rightMovement; j > i; j--) {
                rightMovement = j;
                right = height[j];
                
                maxVolume = max(maxVolume, min(left, right) * (rightMovement - leftMovement));

                    if (left < right) {
                        break;
                    }
                }
            }

            if (leftMovement == rightMovement) {
                break;
            }
        }



        /*
        for (int i = 0; i < height.size()-1; i++) {
            for (int j = i+1; j < height.size(); j++) {
                distance = j - i;
                maxVolume = max(maxVolume, min(height[i], height[j]) * distance);
            }
        }
        */
        return maxVolume;
    }
};

File: 0011-container-with-most-water/0011-container-with-most-water.go
func maxArea(height []int) int {
    ret := 0
    size := len(height)
    r := size-1

    for l := 0; l < size; l++ {

        for ; l < r && height[l] >= height[r]; {
            ret = max(ret, min(height[l], height[r]) * (r-l))
            r--
        }
        
        ret = max(ret, min(height[l], height[r]) * (r-l))

    }

    return ret

}